File: src/main/java/game/Encoder.java
package org.lichess.compression.game;

import java.util.Arrays;
import java.util.Optional;
import java.util.OptionalInt;

import org.lichess.compression.BitReader;
import org.lichess.compression.BitWriter;

public class Encoder {
    private static final ThreadLocal<MoveList> moveList = new ThreadLocal<MoveList>() {
        @Override
        protected MoveList initialValue() {
            return new MoveList();
        }
    };

    private final OpeningTrie openingTrie = OpeningTrie.mostCommonOpenings();

    public byte[] encode(String pgnMoves[]) {
        BitWriter writer = new BitWriter();

        Optional<String> longestCommonOpening = openingTrie.findLongestCommonOpening(pgnMoves);
        longestCommonOpening.ifPresentOrElse(
                opening -> {
                    writer.writeBits(1, 1);
                    writer.writeBits(openingTrie.get(opening), openingTrie.getBitVectorLength());
                },
                () -> writer.writeBits(0, 1));

        long numPliesLongestCommonOpening = longestCommonOpening
                .map(opening -> opening
                        .chars()
                        .filter(c -> c == ' ')
                        .count())
                .orElse(0L);

        Board board = new Board();
        MoveList legals = moveList.get();

        for (int ply = 0; ply < pgnMoves.length; ply++) {
            String pgnMove = pgnMoves[ply];
            // Parse SAN.
            SAN sanMove = SANParser.parse(pgnMove, board);
            if (sanMove == null) return null;

            // Find index in legal moves.
            board.legalMoves(legals);
            legals.sort();
            
            OptionalInt correctIndex = findSanInLegalMoves(sanMove, legals);
            if (correctIndex.isPresent()) {
                int i = correctIndex.getAsInt();
                Move legal = legals.get(i);
                if (ply >= numPliesLongestCommonOpening) Huffman.write(i, writer);
                board.play(legal);
            }
            else {
                return null;
            }
        }

        return writer.toArray();
    }
    
    private OptionalInt findSanInLegalMoves(SAN sanMove, MoveList legals) {
        boolean foundMatch = false;
        int size = legals.size();
        OptionalInt correctIndex = OptionalInt.empty();
        for (int i = 0; i < size; i++) {
            Move legal = legals.get(i);
            if (legal.role == sanMove.role() && legal.to == sanMove.to() && legal.promotion == sanMove.promotion() && Bitboard.contains(sanMove.from(), legal.from)) {
                if (!foundMatch) {
                    // Encode and play.
                    foundMatch = true;
                    correctIndex = OptionalInt.of(i);
                }
                else return OptionalInt.empty();
            }
        }
        return correctIndex;
    }

    public static class DecodeResult {
        public final String pgnMoves[];
        public final Board board;
        public final int halfMoveClock;
        public final byte positionHashes[];
        public final String lastUci;

        public DecodeResult(String pgnMoves[], Board board, int halfMoveClock, byte positionHashes[], String lastUci) {
            this.pgnMoves = pgnMoves;
            this.board = board;
            this.halfMoveClock = halfMoveClock;
            this.positionHashes = positionHashes;
            this.lastUci = lastUci;
        }
    }

    public DecodeResult decode(byte input[], int plies) {
        BitReader reader = new BitReader(input);

        String output[] = new String[plies];
        
        long numPliesDecodedOpening = 0L;

        if (reader.readBits(1) == 1) {
            int code = reader.readBits(openingTrie.getBitVectorLength());
            Optional<String> decodedOpening = openingTrie.getFirstOpeningMappingToCode(code);
            decodedOpening.ifPresent(opening -> {
               String moves[] = opening.split(" ");
               for (int i = 0; i < Math.min(plies, moves.length); i++) {
                   output[i] = moves[i];
               }
            });
            numPliesDecodedOpening = decodedOpening
                    .map(opening -> opening
                            .chars()
                            .filter(c -> c == ' ')
                            .count())
                    .orElse(0L);
        }

        Board board = new Board();
        MoveList legals = moveList.get();

        String lastUci = null;

        // Collect the position hashes (3 bytes each) since the last capture
        // or pawn move.
        int lastZeroingPly = -1;
        int lastIrreversiblePly = -1;
        byte positionHashes[] = new byte[3 * (plies + 1)];
        setHash(positionHashes, -1, board.zobristHash());

        for (int i = 0; i <= plies; i++) {
            if (0 < i || i < plies) board.legalMoves(legals);

            // Append check or checkmate suffix to previous move.
            if (0 < i) {
                if (board.isCheck()) output[i - 1] += (legals.isEmpty() ? "#" : "+");
            }

            // Decode and play next move.
            if (i < plies) {
                legals.sort();
                Move move;
                if (i >= numPliesDecodedOpening) {
                    move = legals.get(Huffman.read(reader));
                    output[i] = move.san(legals);
                }
                else {
                    SAN sanMove = SANParser.parse(output[i], board);
                    int correctIndex = findSanInLegalMoves(sanMove, legals).getAsInt();
                    move = legals.get(correctIndex);
                }
                board.play(move);

                if (move.isZeroing()) lastZeroingPly = i;
                if (move.isIrreversible()) lastIrreversiblePly = i;
                setHash(positionHashes, i, board.zobristHash());

                if (i + 1 == plies) lastUci = move.uci();
            }
        }

        return new DecodeResult(
            output,
            board,
            plies - 1 - lastZeroingPly,
            Arrays.copyOf(positionHashes, 3 * (plies - lastIrreversiblePly)),
            lastUci);
    }

    private static void setHash(byte buffer[], int ply, int hash) {
        // The hash for the starting position (ply = -1) goes last. The most
        // recent position goes first.
        int base = buffer.length - 3 * (ply + 1 + 1);
        buffer[base] = (byte) (hash >>> 16);
        buffer[base + 1] = (byte) (hash >>> 8);
        buffer[base + 2] = (byte) hash;
    }
}

File: src/main/scala/game/Encoder.scala
package org.lichess.compression.game

import org.lichess.compression.BitOps

object Encoder:

  private val moveList = new ThreadLocal[MoveList]:
    override def initialValue(): MoveList = MoveList()

  private val SAN_RE =
    "([NBKRQ])?([a-h])?([1-8])?x?([a-h][1-8])(?:=([NBRQK]))?[\\+#]?".r

  private def charToRole(c: Char): Role = c match
    case 'N' => Role.KNIGHT
    case 'B' => Role.BISHOP
    case 'R' => Role.ROOK
    case 'Q' => Role.QUEEN
    case 'K' => Role.KING
    case _   => throw IllegalArgumentException()

  def encode(pgnMoves: Array[String]): Array[Byte] =
    val writer = BitOps.Writer()
    val board  = Board()
    val legals = moveList.get()

    for pgnMove <- pgnMoves do
      var role: Role      = null
      var promotion: Role = null
      var from: Long      = Bitboard.ALL
      var to: Int         = 0

      if pgnMove.startsWith("O-O-O") then
        role = Role.KING
        from = board.kings
        to = Bitboard.lsb(board.rooks & Bitboard.RANKS(if (board.turn) 0 else 7))
      else if pgnMove.startsWith("O-O") then
        role = Role.KING
        from = board.kings
        to = Bitboard.msb(board.rooks & Bitboard.RANKS(if (board.turn) 0 else 7))
      else
        SAN_RE.findFirstMatchIn(pgnMove) match
          case None =>
            return null
          case Some(matcher) =>
            val roleStr = matcher.group(1)
            role = if (roleStr == null) Role.PAWN else charToRole(roleStr.charAt(0))
            if matcher.group(2) != null then from &= Bitboard.FILES(matcher.group(2).charAt(0) - 'a')
            if matcher.group(3) != null then from &= Bitboard.RANKS(matcher.group(3).charAt(0) - '1')

            to = Square.square(matcher.group(4).charAt(0) - 'a', matcher.group(4).charAt(1) - '1')

            if matcher.group(5) != null then promotion = charToRole(matcher.group(5).charAt(0))

      board.legalMoves(legals)
      legals.sort()
      var foundMatch = false
      for i <- 0 until legals.getSize() do
        val legal = legals.get(i)
        if legal.role == role && legal.to == to && legal.promotion == promotion && Bitboard.contains(
            from,
            legal.from
          )
        then
          if foundMatch then return null
          Huffman.write(i, writer)
          board.play(legal)
          foundMatch = true

      if !foundMatch then return null

    writer.toArray()
  end encode

  case class DecodeResult(
      pgnMoves: Array[String],
      board: Board,
      halfMoveClock: Int,
      positionHashes: Array[Byte],
      lastUci: String
  )

  def decode(input: Array[Byte], plies: Int): DecodeResult =
    val reader              = BitOps.Reader(input)
    val output              = Array.ofDim[String](plies)
    val board               = Board()
    val legals              = moveList.get()
    var lastUci: String     = null
    var lastZeroingPly      = -1
    var lastIrreversiblePly = -1
    val positionHashes      = Array.ofDim[Byte](3 * (plies + 1))
    setHash(positionHashes, -1, board.zobristHash())

    for i <- 0 to plies do
      if 0 < i || i < plies then board.legalMoves(legals)

      if 0 < i then if board.isCheck() then output(i - 1) += (if legals.isEmpty then "#" else "+")

      if i < plies then
        val moveIndex = Huffman.read(reader)
        legals.partialSort(moveIndex + 1)
        val move = legals.get(moveIndex)
        output(i) = san(move, legals)
        board.play(move)

        if move.isZeroing then lastZeroingPly = i
        if move.isIrreversible then lastIrreversiblePly = i
        setHash(positionHashes, i, board.zobristHash())

        if i + 1 == plies then lastUci = move.uci()
    end for

    DecodeResult(
      output,
      board,
      plies - 1 - lastZeroingPly,
      positionHashes.slice(0, 3 * (plies - lastIrreversiblePly)),
      lastUci
    )
  end decode

  private def san(move: Move, legals: MoveList): String = move.`type` match
    case Move.NORMAL | Move.EN_PASSANT =>
      val builder = StringBuilder(6)
      builder.append(move.role.symbol)

      if move.role != Role.PAWN then
        var file   = false
        var rank   = false
        var others = 0L

        for i <- 0 until legals.getSize() do
          val other = legals.get(i)
          if other.role == move.role && other.to == move.to && other.from != move.from then
            others |= 1L << other.from

        if others != 0 then
          if (others & Bitboard.RANKS(Square.rank(move.from))) != 0 then file = true
          if (others & Bitboard.FILES(Square.file(move.from))) != 0 then rank = true
          else file = true

        if file then builder.append((Square.file(move.from) + 'a').toChar)
        if rank then builder.append((Square.rank(move.from) + '1').toChar)
      else if move.capture then builder.append((Square.file(move.from) + 'a').toChar)

      if move.capture then builder.append('x')

      builder.append((Square.file(move.to) + 'a').toChar)
      builder.append((Square.rank(move.to) + '1').toChar)
      if move.promotion != null then builder.append(s"=${move.promotion.symbol}")

      builder.toString()

    case Move.CASTLING =>
      if move.from < move.to then "O-O" else "O-O-O"

    case _ => "--"
  end san

  private def setHash(buffer: Array[Byte], ply: Int, hash: Int): Unit =
    val base = buffer.length - 3 * (ply + 1 + 1)
    buffer(base) = (hash >>> 16).toByte
    buffer(base + 1) = (hash >>> 8).toByte
    buffer(base + 2) = hash.toByte
end Encoder
